### ISSUE TITLE
Addition of Undetermined FASTQs to non-pooled samples

### ISSUE THREAD

**Issue by Vince-janv** [Clinical-Genomics/cg#2419]
## Description

On old flow cells some samples were run alone on a lane (currently samples are always pooled and separated by index). For these samples the undetermined FASTQ file were added to the sample in housekeeper since we could guarantee that the reads came from that sample.  

This logic was handled [here](https://github.com/Clinical-Genomics/demultiplexing/blob/9680e28fc0d406320c863dd98a37d19db3d840db/scripts/hiseqx/xdemuxtiles.batch#L128-L139C11) in the demultiplexing repo.

For these old samples to get all their files when re-demultiplexing we need to implement this logic.
## Suggested solution

TBD

## This can be closed when

Undetermined FASTQ files are added to non-pooled samples

**Comment by henrikstranneheim** [comment]
@seallard Could you work on this? If refinement is needed please coordinate with @Vince-janv and @karlnyr .

**Comment by seallard** [comment]
@Vince-janv @karlnyr Do we want to generate sequencing metrics based on these undetermined fastqs as well? Should the read count for the samples be updated with the undetermined fastqs assigned to it?

**Comment by Vince-janv** [comment]
I've looked trough this and it seems like the following was done previously. If the sample was alone on a lane the metrics assigned to the sample, and the metrics for the undetermined reads were baked into one unaligned entry. 

I fear we might have to do the same. I don't think we can have just add the files to HK without updating the reads, and I don't think we can have the reads not be the sum of the sequencing_metrics entries. I hope this is not that hard to implement 😬

**Comment by seallard** [comment]
The required data seems to be available in the generated metrics files. So it should be doable, might just need to update the sequencing metrics calculator module.

For the sequencing metrics model we need

- Sample Total Reads In Lane
- Sample Base Percentage Passing Q30 
- Sample Base Mean Quality Score

These are calculated from the statistics files generated by the demultiplexing software. They are reported for bclconvert

| Lane | SampleID    | Sample_Project | index | index2 | ReadNumber |      Yield      |  YieldQ30   | QualityScoreSum | Mean Quality Score (PF) | % Q30 |
|------|-------------|----------------|-------|--------|------------|-----------------|-------------|-----------------|------------------------|-------|
| 2    | Undetermined| Undetermined   |       |        |     1      |  2196984528     | 1886452123  | 75448828497     |          34.34         |  0.86 |
| 2    | Undetermined| Undetermined   |       |        |     2      |  2196984528     | 1717999009  | 71912282552     |          32.73         |  0.78 |

For bcl2fastq they are reported like
```json
   "Undetermined" : {
    "NumberReads" : 5293801,
    "Yield" : 1598727902,
    "ReadMetrics" : [
     {
      "ReadNumber" : 1,
      "Yield" : 799363951,
      "YieldQ30" : 719898156,
      "QualityScoreSum" : 30157716143,
      "TrimmedBases" : 0
     },
     {
      "ReadNumber" : 2,
      "Yield" : 799363951,
      "YieldQ30" : 615436475,
      "QualityScoreSum" : 27313363301,
      "TrimmedBases" : 0
     }
    ]
   }
```

**Comment by Vince-janv** [comment]
KN agreed that our suggested solution to merge the undetermined metrics for these samples with the ones we add already.

**Comment by seallard** [comment]
The new logic will look something like

![image](https://github.com/Clinical-Genomics/cg/assets/33596072/5fa5d991-8a30-4f9d-b2a8-f337a8ff3ecd)

**Comment by seallard** [comment]
> > The new logic will look something like
> 
> @seallard I'm guessing the split is based on yes/no? I really like having all this as a separate step after the "normal" part. Well done 👍

Yeah, the split is whether it exists or not already in the metrics table in status db. Thanks 🙂 Still not 100 % how to structure it, pending back and forth. But probably cleaner to separate the handling of undetermined reads completely.

**Comment by seallard** [comment]
The solution outlined above is not idempotent. So we would update existing metrics multiple times if the post processing flow was run multiple times for a flow cell. So the existing flow has to be adjusted slightly so that all metrics are written at once to the database.
